spfx upgrade reducing the number of commands required to install dependencies

baywet opened the pull request:
Adds #508 
Implemented a very simple map reduce to avoid running multiple times the same installation command for different dependencies, thus accelerating install time and upgrade sequence.

waldekmastykarz commented:
Thanks! I'll have a look at it shortly.

waldekmastykarz commented:
Glancing at the code, I wonder if we couldn't simplify the matching piece and use string comparison instead of a regex, given that we control how the install/uninstall are written ourselves.

baywet commented:
What's the concern with a regex? It more "intelligent" does validation (eg: avoids matching -E in a package name and confusing it for a exact save…). Maintainability? Performance?

waldekmastykarz commented:
Primarily maintainability plus the nested if constructs used to go through the different cases. It would be good if we could simplify the code

baywet commented:
I couldn't use a map structure because of the existing for each. And I didn't want to change too much code.
I can extract that to a method and use constants to improve readability if you think that's valuable?

waldekmastykarz commented:
Not sure I fully understand the difference. If not too much work, could you post the altered code snippet here and we can then decide which one would be easier to maintain over time?

baywet commented:
moved the code to these two methods, which makes readability much better. Added constants for the regex capture groups, this way it's easier to avoid mistakes maintaining and helps with reading as well.
https://gist.github.com/baywet/fb1dacca361a05568fcebfd162e86f9a

waldekmastykarz commented:
I think I prefer the original approach plus the constants. I don't see the separate functions adding as much.
We could simplify the regex because we either use `-SE` or `-DE`, there is no separate `-S` or `-D`.

baywet commented:
having separate methods makes the getReportMd method easier to read IMHO. (one scroll rule ;-) )
I'll push new commits ASAP

waldekmastykarz commented:
Ah, that's likely easier to see when integrated with the rest of the code. Would functions get easier too if we simplified the regex and dropped some of the arrays?

baywet commented:
Well if we say the rules will always input SE or DE yes (making that modification as we "speak") it eliminates some cases.

waldekmastykarz commented:
That's what we currently have and I don't see at the moment why we'd need other cases. So `i -DE`, `i -SE` and `un`

baywet commented:
for uninstall, shouldn't we add -S or -D in the initial command (and keep the difference in the map reduce)? this way it's saved to the package.json even with the older versions of npm?  https://docs.npmjs.com/cli/uninstall

waldekmastykarz commented:
For the time being that might be a safer choice, yes 👍

baywet commented:
@waldekmastykarz this is by design if you look at this line https://github.com/baywet/office365-cli/blob/feature/npmimapreduce/src/o365/spfx/commands/project/project-upgrade.ts#L331. The reason I did that it to avoid people doing installation line by line and notice at the end "oh there's a line that does everything at once". It also makes the report smaller, thus easier to read and understanding what needs to be done.

waldekmastykarz commented:
Sorry, but we can't provide an incomplete upgrade report. The first part of the report is meant to help people understand the changes. The summary is meant for when you know what each change is for and just need to grab the commands to run the update. In both cases, the information should be complete.

baywet commented:
What if I push a resolution per condensate command like this is done here then? https://github.com/baywet/office365-cli/blob/1203efc3e5f587246d14129fa8e8aeacb108a3c5/src/o365/spfx/commands/project/project-upgrade.ts#L348
You'd effectively have "execute the following command: npm i module1 module2..."

waldekmastykarz commented:
Sorry, I don't think I understand what you mean

waldekmastykarz commented:
@baywet I'd reserve that for the summary. In the first section of the report, I'll prefer to have each finding separately so that if necessary, we can include additional information about the finding, like information about the package, its purpose, etc. If you don't need all that information and are just interested in the scripts, you can grab them from the summary section.

baywet commented:
here is an update, let me know what you think

waldekmastykarz commented:
Thanks for the quick fix. I'll have a look shortly 👍

waldekmastykarz commented:
Merged manually. Thanks again!
